### ISSUE TITLE
why is n_negative_keep necessary

### ISSUE THREAD

**Issue by chenzhuo8804** [pierluigiferrari/ssd_keras#209]
I am not very sure why **n_negative_keep** is needed in line **166** of keras_ssd_loss.py, as an upper bounding of how many losses will be counted, documentation says that tf.nn.top_k will  possibly return indices of positive boxes.

For example we received a vector of all zeros as loss for negative boxes, and choose the top-k, for example, 3, losses, my understanding is that top-k will just return [0, 0, 0] as loss value,

Please correct me if I am wrong

Thanks!

**Comment by pierluigiferrari** [comment]
The implementation is correct as is, but it could certainly be implemented in a way so that you wouldn't need `n_negative_keep`. Without having thought this through in detail, in line 190, you could probably replace

`neg_class_loss = tf.reduce_sum(classification_loss * negatives_keep, axis=-1)`

by

`neg_class_loss = tf.reduce_sum(neg_class_loss_all * negatives_keep, axis=-1)`

and then you wouldn't need `n_negative_keep`. I haven't thought this through in more detail though. However, computing `n_negative_keep` is cheap and the rest of the code would be more or less the same. So I agree that `n_negative_keep` could be done away with, but I don't see any functional or performance disadvantages of the current implementation.

The current implementation definitely needs `n_negative_keep`, albeit just to cover an unlikely edge case.

**Comment by DonghoonPark12** [comment]
@pierluigiferrari 
"If we use `neg_class_loss_all` instead of `classification_loss` then we wouldn't need `n_negative_keep`."

What do you mean??

Isn't `n_negative_keep` needed to make masking of `negative_keep` ??

And I agree that 
`neg_class_loss = tf.reduce_sum(classification_loss * negatives_keep, axis=-1)` 
and
`neg_class_loss = tf.reduce_sum(neg_class_loss_all * negatives_keep, axis=-1)`
is same.
